### PR TITLE
CSHARP-3210: Support Azure and GCP keystores in FLE.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
related libmongocrypt PR: https://github.com/DmitryLukyanov/libmongocrypt/pull/4

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs [hunk 1]
```diff
@@ -65,24 +65,21 @@ private CryptClient CreateCryptClient(CryptOptions options)
...
             if (_kmsProviders != null && _kmsProviders.Count > 0)
             {
-                kmsProvidersMap = new Dictionary<KmsType, IKmsCredentials>();
-                if (_kmsProviders.TryGetValue("aws", out var awsProvider))
+                kmsProvidersMap = new Dictionary<KmsType, KmsCredentials>();
+                foreach (var kmsProvider in _kmsProviders)
                 {
-                    if (awsProvider.TryGetValue("accessKeyId", out var accessKeyId) &&
-                        awsProvider.TryGetValue("secretAccessKey", out var secretAccessKey))
+                    if (!Enum.TryParse<KmsType>(kmsProvider.Key, true, out var kmsType))
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs]: If we decide to defer all validation on the mongocrypt level, then this validation can be removed

File: src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs [hunk 2]
```diff
@@ -65,24 +65,21 @@ private CryptClient CreateCryptClient(CryptOptions options)
...
-                {
-                    if (localProvider.TryGetValue("key", out var keyObject) && keyObject is byte[] key)
-                    {
-                        kmsProvidersMap.Add(KmsType.Local, new LocalKmsCredentials(key));
+                        var message = $"The provided kms provider {kmsProvider.Key} is not supported. The list of supported providers: {string.Join(", ", Enum.GetValues(typeof(KmsType)))}.";
+                        throw new NotSupportedException(message);
                     }
+                    var kmsTypeDocumentKey = kmsProvider.Key.ToLower();
+                    var kmsProviderDocument = CreateProviderDocument(kmsTypeDocumentKey, kmsProvider.Value);
+                    var kmsCredentials = new KmsCredentials(kmsType, credentialsBytes: kmsProviderDocument.ToBson());
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs]: kmsType argument is not required here. The only reason why I left it as is improving developer experience. To avoid the need to convert bytes into something understandable mostly in debugging

File: src/MongoDB.Driver/Encryption/ExplicitEncryptionLibMongoCryptController.cs [hunk 3]
```diff
@@ -219,26 +218,25 @@ public async Task<BsonValue> DecryptFieldAsync(BsonBinaryData wrappedBinaryValue
...
-                    throw new ArgumentException($"Invalid kmsProvider {kmsProvider}.");
+                var message = $"The provided kms provider {kmsProvider} is not supported. The list of supported providers: {string.Join(", ", Enum.GetValues(typeof(KmsType)))}.";
+                throw new NotSupportedException(message);
             }
+            var dataKeyDocument = new BsonDocument("provider", kmsProvider.ToLower());
+            if (masterKey != null)
+            {
+                dataKeyDocument.AddRange(masterKey.Elements);
+            }
+            return new KmsKeyId(kmsType, dataKeyDocument.ToBson(), wrappedAlternateKeyNamesBytes);
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Encryption/ExplicitEncryptionLibMongoCryptController.cs]: kmsType argument is not required here. The only reason why I left it as is - improving developer experience. To avoid the need to convert bytes into something understandable mostly in debugging

File: src/MongoDB.Driver/Encryption/ExplicitEncryptionLibMongoCryptController.cs [hunk 4]
```diff
@@ -219,26 +218,25 @@ public async Task<BsonValue> DecryptFieldAsync(BsonBinaryData wrappedBinaryValue
...
+        private KmsKeyId GetKmsKeyId(string kmsProvider, IReadOnlyList<string> alternateKeyNames, BsonDocument masterKey)
         {
             IEnumerable<byte[]> wrappedAlternateKeyNamesBytes = null;
             if (alternateKeyNames != null)
             {
                 wrappedAlternateKeyNamesBytes = alternateKeyNames.Select(GetWrappedAlternateKeyNameBytes);
             }
 
-            switch (kmsProvider)
+            if (!Enum.TryParse<KmsType>(kmsProvider, true, out var kmsType))
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Encryption/ExplicitEncryptionLibMongoCryptController.cs]: If we decide to defer all validation on the mongocrypt level, then this validation can be removed

File: src/MongoDB.Driver/Encryption/LibMongoCryptControllerBase.cs [hunk 5]
```diff
@@ -235,9 +235,9 @@ private void SendKmsRequest(KmsRequest request, CancellationToken cancellation)
                 var requestBytes = request.Message.ToArray();
                 sslStream.Write(requestBytes);
 
-                var buffer = new byte[4096];
                 while (request.BytesNeeded > 0)
                 {
+                    var buffer = new byte[request.BytesNeeded]; // BytesNeeded is the maximum number of bytes that libmongocrypt wants to receive.
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/Encryption/LibMongoCryptControllerBase.cs:240]: Fixing an old bug. This will cause a problem each time when a kms response is bigger than 1024 bytes. Previously we didn't have such big responses, but now, the azure returns more than 2000 bytes in response.

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientSideEncryptionSmokeProseTests.cs [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+using MongoDB.Driver.Core.Configuration;
+using MongoDB.Driver.Encryption;
+using MongoDB.Driver.TestHelpers;
+using MongoDB.Libmongocrypt;
+using Xunit;
+
+namespace MongoDB.Driver.Tests.Specifications.client_side_encryption.prose_tests
+{
+    // TODO: move into ClientEncryptionProseTests.cs
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientSideEncryptionSmokeProseTests.cs]: should be moved into ClientEncryptionProseTests as soon as the spec will be ready

**Review by DmitryLukyanov — COMMENTED** [review]
https://docs.google.com/document/d/18rO1QYORNvKs8rYWtXKdrT-2YaORMJ-r8xcM8g1SJ8Q/edit

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientSideEncryptionSmokeProseTests.cs (hunk 6 above)
Comment (jyemin) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientSideEncryptionSmokeProseTests.cs]: Kevin just opened a PR on the specs repo with YAML/JSON based tests.  Can this be replaced with those?

Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientSideEncryptionSmokeProseTests.cs]: sure, will do it today

File: src/MongoDB.Driver/Encryption/LibMongoCryptControllerBase.cs (hunk 5 above)
Comment (jyemin) [reply src/MongoDB.Driver/Encryption/LibMongoCryptControllerBase.cs:240]: I'm not sure I understand.  The old buffer was 4096 bytes, so why would 2000 bytes cause a problem?  Also, wouldn't the existing code just continue to feed the response bytes until it was done?

File: src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs [hunk 7]
```diff
@@ -105,7 +97,18 @@ private CryptOptions CreateCryptOptions()
...
-            return new CryptOptions(kmsProvidersMap, schemaBytes);
+            return new CryptOptions(kmsProviders, schemaBytes);
+        }
+
+        private BsonDocument CreateProviderDocument(string kmsType, IReadOnlyDictionary<string, object> data)
+        {
+            var providerContent = new BsonDocument();
+            foreach (var record in data)
+            {
+                providerContent.Add(new BsonElement(record.Key, BsonValue.Create(record.Value)));
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs:108]: Calling `BsonValue.Create` with an `object` argument is dangerous.

Do you know what values for `record.Value` might appear here?

File: tests/MongoDB.Driver.Core.TestHelpers/CoreExceptionHelper.cs [hunk 8]
```diff
@@ -100,6 +100,9 @@ public static Exception CreateException(string errorType)
                 case "TimedOutSocketException":
                     return new SocketException((int)SocketError.TimedOut);
 
+                case "ConnectionRefusedSocketException":
+                    return new SocketException((int)SocketError.ConnectionRefused);
+
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.TestHelpers/CoreExceptionHelper.cs:105]: Is this related to this ticket?

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 9]
```diff
@@ -526,88 +540,151 @@ EncryptionAlgorithm ParseAlgorithm(string algorithm)
...
-            {
-                var testCaseMasterKey = new BsonDocument
-                {
-                    { "region", "us-east-1" },
-                    { "key", "arn:aws:kms:us-east-1:579766882180:key/89fcc2c4-08b0-4bd9-9f25-e30687b580d0" }
-                };
-                TestCase(testCaseMasterKey);
-
-                testCaseMasterKey = new BsonDocument
+            Action<string, Dictionary<string, object>> setValidKmsEndpointFunc =
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: Why are these not local functions?

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver/Encryption/LibMongoCryptControllerBase.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/Encryption/LibMongoCryptControllerBase.cs:240]: it's related to behavior in `mongocrypt.dll`. We discussed this with Kevin https://mongodb.slack.com/archives/C016YF807JS/p1602889908019400 and offline. The buffer size which we pass into `libmongocrypt` is important. And in case if we pass more than 2000 bytes (and more than libmongocrypt expects), it affects inner library logic, I don't remember what was exactly wrong in `c` library, we can discuss it with him

File: tests/MongoDB.Driver.Core.TestHelpers/CoreExceptionHelper.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.TestHelpers/CoreExceptionHelper.cs:105]: it's  related to this test: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/157/files#diff-0537777d32b586e768ac1e74c10677b06f8933c41fba72c3f11a6f74862925e7R547 (CustomEndpointTest)

File: src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs (hunk 7 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs:108]: any which was provided by user, but we expect only `byte[]` and `string`

File: src/MongoDB.Driver/Encryption/LibMongoCryptControllerBase.cs (hunk 5 above)
Comment (kevinAlbs) [reply src/MongoDB.Driver/Encryption/LibMongoCryptControllerBase.cs:240]: The old buffer had a capacity of 4096 bytes, but was not being completely filled when only AWS responses were being returned. Azure returns larger responses. libmongocrypt explicitly errors if it is fed more than the number of bytes it requests, which is documented in the API: https://github.com/mongodb/libmongocrypt/blob/7a53c00f8c80a672c5c6b417d4418da0cccb2acb/src/mongocrypt.h.in#L908:L909.

The initial amount it requests is 1024 bytes (which is arbitrary). And in discussing with @DmitryLukyanov, I think we can remove the arbitrary limit and error in libmongocrypt. I filed https://jira.mongodb.org/browse/MONGOCRYPT-286 to do so. But it seemed more safer to me to have C# follow the current prescribed API now rather than change libmongocrypt's behavior.

File: src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs (hunk 7 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs:108]: discussed offline

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: because I pass these functions as argument here https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/157/files#diff-0537777d32b586e768ac1e74c10677b06f8933c41fba72c3f11a6f74862925e7R595-R596 (line 595-596)

Comment (rstam) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: I understand that you pass them as arguments below.

My question is why they aren't a local function? That wouldn't prevent passing them would it?

Comment (rstam) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: It's not a big deal... it just seemed odd that you have some `Action` variables that looked like they could/should be local functions.

**Review by rstam — CHANGES_REQUESTED** [review]
Not really requesting changes I guess... just continuing a discussion.

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: I didn't know that local functions can be used in this way. Thanks for your suggestion. Done

File: src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj [hunk 10]
```diff
@@ -116,7 +116,7 @@
   <ItemGroup>
     <PackageReference Include="DnsClient" Version="1.3.1" />
     <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.6.2" PrivateAssets="All" />
-    <PackageReference Include="MongoDB.Libmongocrypt" Version="1.0.0" />
+    <PackageReference Include="MongoDB.Libmongocrypt" Version="0.0.0-local" />
```
Comment (rstam) [inline src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj]: I assume you will fix this before pushing to master.

File: src/MongoDB.Driver/MongoDB.Driver.csproj [hunk 11]
```diff
@@ -42,7 +42,7 @@
   </PropertyGroup>
   <ItemGroup>
     <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.6.2" PrivateAssets="All" />
-    <PackageReference Include="MongoDB.Libmongocrypt" Version="1.0.0" />
+    <PackageReference Include="MongoDB.Libmongocrypt" Version="0.0.0-local" />
```
Comment (rstam) [inline src/MongoDB.Driver/MongoDB.Driver.csproj]: I assume you will fix this before pushing to master.

**Review by rstam — APPROVED** [review]
LGTM with the caveat that my review is superficial due to limited knowledge of the code being changed.

File: src/MongoDB.Driver/MongoDB.Driver.csproj (hunk 11 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoDB.Driver.csproj]: Correct

File: src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj (hunk 10 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj]: Correct

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/ClientSideEncryptionTestRunner.cs [hunk 12]
```diff
@@ -207,6 +207,11 @@ private AutoEncryptionOptions ConfigureAutoEncryptionOptions(BsonDocument autoEn
             return autoEncryptionOptions;
         }
 
+        private string GetEnvironmentVariableOrDefaultOrThrowIfNothing(string variableName, string defaultValue = null) =>
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/ClientSideEncryptionTestRunner.cs:210]: Nit: consider renaming to `GetEnvironmentVariableOrDefaultValueOrThrow`

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 13]
```diff
@@ -18,6 +18,7 @@
 using System.Collections.ObjectModel;
 using System.Linq;
 using System.Net.Sockets;
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: Nit: unnecessary using

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 14]
```diff
@@ -526,91 +540,154 @@ EncryptionAlgorithm ParseAlgorithm(string algorithm)
...
-                exception.Message.Should().Contain("parse error");
+            Guid CreateDataKeyTestCaseStep(ClientEncryption testCaseClientEncription, BsonDocument masterKey, bool async)
+            {
+                var dataKeyOptions = new DataKeyOptions(masterKey: masterKey);
+                return CreateDataKey(testCaseClientEncription, kmsType, dataKeyOptions, async);
+            }
 
-                // additional not spec tests
-                testCaseMasterKey = new BsonDocument
+            void SetInvalidKmsEndpointFunc(string kt, Dictionary<string, object> ko)
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` -> `Action`. Or maybe it makes sense to leave this method without postfix.

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 15]
```diff
@@ -526,91 +540,154 @@ EncryptionAlgorithm ParseAlgorithm(string algorithm)
...
+                        ko.Add("identityPlatformEndpoint", "example.com:443");
+                        break;
+                    case "gcp":
+                        ko.Add("endpoint", "example.com:443");
+                        break;
+                }
+            }
 
-                void TestCase(BsonDocument masterKey)
+            void SetValidKmsEndpointFunc(string kt, Dictionary<string, object> ko)
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` -> `Action`. Or maybe it makes sense to leave this method without postfix.

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 16]
```diff
@@ -526,91 +540,154 @@ EncryptionAlgorithm ParseAlgorithm(string algorithm)
...
+            string kmsType,
+            string customEndpoint,
+            string expectedExceptionInfoForValidEncryption,
+            string expectedExceptionInfoForInvalidEncryption)
         {
             RequireServer.Check().Supports(Feature.ClientSideEncryption);
 
             using (var client = ConfigureClient())
-            using (var clientEncryption = ConfigureClientEncryption(client.Wrapped as MongoClient))
+            using (var clientEncryption = ConfigureClientEncryption(client.Wrapped as MongoClient, SetValidKmsEndpointFunc))
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` is usually a delegate which returns a value. In this case this is `Action`. I would suggest to keep the name consistent here and in all of the occurrences.

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 17]
```diff
@@ -526,91 +540,154 @@ EncryptionAlgorithm ParseAlgorithm(string algorithm)
...
+            using (var clientEncryptionInvalid = ConfigureClientEncryption(client.Wrapped as MongoClient, SetInvalidKmsEndpointFunc))
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` -> `Action`

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 18]
```diff
@@ -728,12 +805,12 @@ private DisposableMongoClient ConfigureClient(bool clearCollections = true)
             return clientEncrypted;
         }
 
-        private ClientEncryption ConfigureClientEncryption(MongoClient client)
+        private ClientEncryption ConfigureClientEncryption(MongoClient client, Action<string, Dictionary<string, object>> updateKmsProviderFunc = null)
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` -> `Action`

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 19]
```diff
@@ -728,12 +805,12 @@ private DisposableMongoClient ConfigureClient(bool clearCollections = true)
...
         {
             var clientEncryptionOptions = new ClientEncryptionOptions(
                 keyVaultClient: client.Settings.AutoEncryptionOptions?.KeyVaultClient ?? client,
                 keyVaultNamespace: __keyVaultCollectionNamespace,
-                kmsProviders: GetKmsProviders());
+                kmsProviders: GetKmsProviders(updateKmsProviderFunc));
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` -> `Action`

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 20]
```diff
@@ -940,27 +1037,55 @@ private IMongoCollection<BsonDocument> GetCollection(IMongoClient client, Collec
...
         }
 
-        private IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> GetKmsProviders()
+        private string GetEnvironmentVariableOrDefaultOrThrowIfNothing(string variableName, string defaultValue = null) =>
+            Environment.GetEnvironmentVariable(variableName) ??
+            defaultValue ??
+            throw new Exception($"{variableName} system variable should be configured on the machine.");
+
+
+        private IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> GetKmsProviders(Action<string, Dictionary<string, object>> updateKmsProviderFunc = null)
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` -> `Action`

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 21]
```diff
@@ -940,27 +1037,55 @@ private IMongoCollection<BsonDocument> GetCollection(IMongoClient client, Collec
...
+            var awsAccessKey = GetEnvironmentVariableOrDefaultOrThrowIfNothing("FLE_AWS_ACCESS_KEY_ID");
+            var awsSecretAccessKey = GetEnvironmentVariableOrDefaultOrThrowIfNothing("FLE_AWS_SECRET_ACCESS_KEY");
+            var awsKmsOptions = new Dictionary<string, object>
             {
-                { "region", awsRegion },
                 { "accessKeyId", awsAccessKey },
                 { "secretAccessKey", awsSecretAccessKey }
             };
-            kmsProviders.Add("aws", kmsOptions);
+            updateKmsProviderFunc?.Invoke("aws", awsKmsOptions);
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` -> `Action`

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 22]
```diff
@@ -940,27 +1037,55 @@ private IMongoCollection<BsonDocument> GetCollection(IMongoClient client, Collec
...
+            kmsProviders.Add("aws", awsKmsOptions);
 
             var localOptions = new Dictionary<string, object>
             {
                 { "key", new BsonBinaryData(Convert.FromBase64String(LocalMasterKey)).Bytes }
             };
+            updateKmsProviderFunc?.Invoke("local", localOptions);
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` -> `Action`

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 23]
```diff
@@ -940,27 +1037,55 @@ private IMongoCollection<BsonDocument> GetCollection(IMongoClient client, Collec
...
+            var azureTenantId = GetEnvironmentVariableOrDefaultOrThrowIfNothing("FLE_AZURE_TENANT_ID");
+            var azureClientId = GetEnvironmentVariableOrDefaultOrThrowIfNothing("FLE_AZURE_CLIENT_ID");
+            var azureClientSecret = GetEnvironmentVariableOrDefaultOrThrowIfNothing("FLE_AZURE_CLIENT_SECRET");
+            var azureKmsOptions = new Dictionary<string, object>
+            {
+                { "tenantId", azureTenantId },
+                { "clientId", azureClientId },
+                { "clientSecret", azureClientSecret }
+            };
+            updateKmsProviderFunc?.Invoke("azure", azureKmsOptions);
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` -> `Action`

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 24]
```diff
@@ -940,27 +1037,55 @@ private IMongoCollection<BsonDocument> GetCollection(IMongoClient client, Collec
...
+            kmsProviders.Add("azure", azureKmsOptions);
+
+            var gcpEmail = GetEnvironmentVariableOrDefaultOrThrowIfNothing("FLE_GCP_EMAIL");
+            var gcpPrivateKey = GetEnvironmentVariableOrDefaultOrThrowIfNothing("FLE_GCP_PRIVATE_KEY");
+            var gcpKmsOptions = new Dictionary<string, object>
+            {
+                { "email", gcpEmail },
+                { "privateKey", gcpPrivateKey }
+            };
+            updateKmsProviderFunc?.Invoke("gcp", gcpKmsOptions);
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: `Func` -> `Action`

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 25]
```diff
@@ -940,27 +1037,55 @@ private IMongoCollection<BsonDocument> GetCollection(IMongoClient client, Collec
                 .GetCollection<BsonDocument>(collectionNamespace.CollectionName, collectionSettings);
         }
 
-        private IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> GetKmsProviders()
+        private string GetEnvironmentVariableOrDefaultOrThrowIfNothing(string variableName, string defaultValue = null) =>
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:1039]: Nit: consider renaming to `GetEnvironmentVariableOrDefaultValueOrThrow`

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 26]
```diff
@@ -940,27 +1037,55 @@ private IMongoCollection<BsonDocument> GetCollection(IMongoClient client, Collec
...
+            Environment.GetEnvironmentVariable(variableName) ??
+            defaultValue ??
+            throw new Exception($"{variableName} system variable should be configured on the machine.");
+
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:1043]: Remove blank line

**Review by MikalaiMazurenka — APPROVED** [review]
Great job! LGTM with a few minor comments which are mostly up to you.

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/ClientSideEncryptionTestRunner.cs (hunk 12 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/ClientSideEncryptionTestRunner.cs:210]: I don't see a significant difference but the current way looks more clear regarding what happens in the throw step. So since it's a nit I would leave it as is

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 13 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: done

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 16 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: We use `Configurator` postfix in the MongoClientSettings. So I took this template

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 17 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: see above

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 14 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: see above

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 15 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: see above

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 19 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: see above

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 18 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: see above

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 21 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: see above

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 22 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: see above

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 23 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: see above

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 24 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: see above

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 26 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:1043]: done

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 16 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver/MongoClientSettings.cs#L195
